platforms/suspend.c

```diff
@@ -44,8 +46,22 @@ bool suspend_wakeup_condition(void) {
     matrix_power_up();
     matrix_scan();
     matrix_power_down();
+
+    bool wakeup = false;
     for (uint8_t r = 0; r < MATRIX_ROWS; r++) {
-        if (matrix_get_row(r)) return true;
+        wakeup_matrix[r] = matrix_get_row(r);
+        wakeup |= wakeup_matrix[r] != 0;
+    }
+
+    return wakeup;
+}
+
+bool keypress_is_wakeup_key(uint8_t row, uint8_t col) {
+    return (wakeup_matrix[row] & ((matrix_row_t)1 << col));
+}
+
+void process_wakeup_matrix(uint8_t row, uint8_t col, bool pressed) {
+    if (!pressed) {
+        wakeup_matrix[row] &= ~((matrix_row_t)1 << col);
     }
```

Comment:
You're right. I have added a `update_wakeup_matrix` function in 73cdf4198652617976e70b645b0d9a27c0f5ac76 which is called after the wake up delay and resets only the keys which have been released during the suspend. This should mitigate the issue?

Reply:
@sigprof friendly ping :slightly_smiling_face:

Reply:
@sigprof Please have another look. I'm sure this addresses the issue and I want to resolve the conversation if possible.

Reply:
That `update_wakeup_matrix()` call may mitigate the problem in some cases, but still leaves some possibilities for a wrong behavior (e.g., if the debouncing algorithm still returns the key state as pressed during `update_wakeup_matrix()`, but reports the key as released after the next `matrix_scan()` call from `matrix_task()`).

Doing things in a 100% correct way could be tricky though:
- Technically the host can suspend the keyboard even while some keys are held pressed (even though the keyboard would immediately send a remote wakeup request, although the host may disable remote wakeup); in this case the pressed state of such keys must not be lost, so that the eventual key release would be handled and won't leave a stuck key or some other stuck state.
- Generating key release events without the corresponding key press events would be bad too, so just setting bits in `matrix_previous` for the keys which were pressed during the wakeup won't work — the extra `wakeup_matrix` array would still be needed.

So the logic should be:
1. Key not pressed before suspend, and not pressed after wakeup → do nothing (normal case).
2. Key not pressed before suspend, but pressed after wakeup → set the `wakeup_matrix` bit to 1 (so that the press and release events would be suppressed).
3. Key pressed before suspend, but not pressed after wakeup → do nothing (the release event will be generated on the first `matrix_task()` call after the wakeup).
4. Key pressed before suspend, and still pressed after wakeup → do nothing (the release event will be generated some time later).

The suggested implementation does not distinguish the cases 2 and 4, and therefore would mishandle the case 4 (even though it would probably be a rare one), and also would mishandle the case 2 which appears as 1 to `matrix_task()` (because the debouncing algorithm stops reporting the pressed state at just the right time).

A fully correct logic would need to have access to `matrix_previous` and would handle the cases above as follows:
1. `matrix_previous` bit is 0, `matrix_current` bit is 0 → do nothing (normal case);
2. `matrix_previous` bit is 0, `matrix_current` bit is 1 → set the `matrix_previous` bit to 1 (this will suppress the press event), set the `wakeup_matrix` bit to 1 (this will suppress the subsequent release event).
3. `matrix_previous` bit is 1, `matrix_current` bit is 0 → do nothing (`matrix_task()` will generate the release event if the matrix state stays like that, or won't do anything if the key gets reported as pressed again — so a momentary release of the key may be missed, but there would be no press/release state inconsistencies).
4. `matrix_previous` bit is 1, `matrix_current` bit is 1 → do nothing (the key release at some point in the future needs to be handled normally).
So the function to update the matrix state after the wakeup would be like this:
```c
void update_matrix_state_after_wakeup(void) {
    for (uint8_t r = 0; r < MATRIX_ROWS; r++) {
        const matrix_row_t current_row = matrix_get_row(row);
        wakeup_matrix[row] |= current_row & ~matrix_previous[row];
        matrix_previous[row] |= current_row;
    }
}
```